### DISCUSSION TITLE
Selecting the closest module from the SAM-list

### DISCUSSION THREAD

**Discussion by casaout** [pvlib/pvlib-python#1344] [Q&A]
I am trying to load the module info for two specific modules using [pvlib.pvsystem.retrieve_sam()](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.pvsystem.retrieve_sam.html). The function loads the lists from [SAM](https://github.com/NREL/SAM/tree/develop/deploy/libraries). 

Unfortunately, these two modules are not available in these lists, so I am trying to identify the most similar ones.

I am looking for these:
* SunPower SPR-333NE-WHT
* SunPower_SPR_400E_WHT_D

What are recommendations for picking the closest modules to the ones available in the list?
Or, is it recommended to create a PR to the SAM-list and adding the modules I am looking for?

**Comment by cwhanse** [comment]
> Unfortunately, these two modules are not available in these lists, so I am trying to identify the most similar ones.
> 
> I am looking for these:
> 
> * SunPower SPR-333NE-WHT
> * SunPower_SPR_400E_WHT_D
> 
> What are recommendations for picking the closest modules to the ones available in the list?

I would pick a module with the same model and power rating closest to the one you want. If you are after CEC model parameters, you can scale the IL_ref parameter by the ratio of powers and the IV curves should be close.

> Or, is it recommended to create a PR to the SAM-list and adding the modules I am looking for?

I don't think so. The SAM list is generated by putting the CEC list through some processing.

The SAM module list that pvlib-python distributes is usually behind the latest SAM release. In this case, both modules you want are in the latest [SAM release](https://github.com/NREL/SAM/blob/develop/deploy/libraries/CEC%20Modules.csv).  Look for 'SunPower SPR-E20-333' and 'SunPower SPR-400E-WHT-D'

**Comment by casaout** [comment]
Dear @cwhanse 

Thank you for your quick response! Where do I find out what exactly each column means? 
To determine the right columns to consider/compare, I am assuming this would be `index`, `STC`,  and `I_L_ref`, or any other ones? How could I scale `I_L_ref`?

Whops, for the second module, I accidentally listed the wrong one, correct would be `SunPower SPR-MAX3-400` (for now I've picked `SunPower SPR-400E-WHT-D`.

What would be the recommended way to use the latest SAM release via pvlib-python?

Thank you

**Reply by cwhanse** [reply]
> Where do I find out what exactly each column means?

Either the docstring for [calcparams_cec](https://github.com/pvlib/pvlib-python/blob/aba071f707f9025882e57f3e55cc9e3e90e869b2/pvlib/pvsystem.py#L2021) or Section 9.3 of the [SAM reference](https://www.nrel.gov/docs/fy15osti/64102.pdf)

To scale one module, you want to be sure that the two modules have the same number of cells in series, and similar Voc and Vmp. The higher power module is made of cells that are more efficient, but otherwise the same. Manufacturers sort cells into efficiency bins to produce e.g. 375W, 385W and 400W modules from the same cell production.

To scale, multiple the I_L_ref of the available module by the ratio of power at STC. For example the I_L_ref parameter for a 400W module, using a 370W module with I_L_ref of 6.9A, would be 6.9A * (400 / 370) =  7.46A.

You should be able to download the latest SAM file (at the link above) and pass the path + name of the file to the `path` parameter of `retrieve_sam`, leaving the `name` parameter as `None`. Or, you can provide the URL to the SAM file to the `path` parameter again leaving `name=None`.

The `retrieve_sam` function could use an overhaul to make it easier to read custom files. Current behavior is to read the specific files distributed with pvlib-python if the user provides any value to `name`.

**Reply by kandersolar** [reply]
Here's another page useful for understanding column names: https://pvpmc.sandia.gov/PVLIB_Matlab_Help/html/pvl_SAMLibraryReader_CECModules_help.html

**Reply by casaout** [reply]
Thank you very much, this helped a lot. 

I can now successfully load the cec-list directly from the Gitrepo. In case someone else is interested, my code:
```
url = "https://raw.githubusercontent.com/NREL/SAM/develop/deploy/libraries/CEC%20Modules.csv"
cec_modules = pvsystem.retrieve_sam(name=None, path=url)
cec_inverters = pvsystem.retrieve_sam('cecinverter')
```

I now found the module with the closest Voc and Vmp, interestingly no module in the list has the same number of cells (104 vs 128). Both have the same STC. This would suggest I need to scale according to the number of cells, right. How exactly should I proceed in that case and which columns should I scale?

| | SunPower SPR-MAX3-400 (target) | SunPower SPR-400E-WHT-D (closest) | scaled (i.e. /128*104)|
|--|--|--|--|
|number cells | 104 | 128 | |
|v_oc_ref|75.6V|85.3V|69.3V|
|v_mp_ref|65.8V|72.9V|59.2V|
|STC|400W|400.221W|no scaling?|
|I_L_ref|??|5.87477A|?|

Is there a reason not to _manually_ enter all the values of the module, instead of adapting an item from the cec_modules list?

Thank you.

**Reply by cwhanse** [reply]
Whether the SPR-400E-WHT-D module is a good surrogate for the SPR-MAX3-400, depends on what you expect from the performance model. If you are interested only in power, I think the 400E module will be close. But if power is the only outcome of interest, perhaps you could use the PVWatts model options and bypass the need to find a surrogate.

An alternative is to get the datasheet for the SPR-MAX3-400 and use pvlib.ivtools.sdm.fit_cec_sam to get the CEC model parameters. You have to install NREL-PySAM, and the algorithm usually converges, but there's no guarantee.

**Reply by casaout** [reply]
Thank you. 

My interests lie in running the [ModelChain ](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.modelchain.ModelChain.html) and using the predicted `mc.ac` (power), as well as the `mc.effective_irradiance` and `mc.total_irrad`.

Assuming that I'll follow the `scaling the model` route, you mentioned that the number of cells, Voc and Vmp should be considered. While Voc and Vmp is similar, the number of cells is not; and I haven't found another module that also has 104 cells in the `CEC%20Modules.csv`. How would I scale in that case? Ultimately, I am forwarding the module parameters into [`pvsystem.PVSystem`](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.pvsystem.PVSystem.html).

Going the [`pvlib.ivtools.sdm.fit_cec_sam`](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.ivtools.sdm.fit_cec_sam.html) route, I could basically also forward the resulting estimated module parameters into `pvsystem.PVSystem`, I assume?

**Reply by cwhanse** [reply]
On the "scaling" route: you can try scaling the voltage per cell to the number of cells in the SPR-MAX3-400 module, but then I think you'll end up in a balancing act trying to get back to 400W STC power. I can't see any of the Maxeon modules in the CEC list, suggesting that they aren't sold in the U.S. for some reason. The output of `fit_cec_sam` (a tuple of parameter values) is what is needed for `pvlib.pvsystem.calcparams_cec` but you will need to code around the differences in data type and ordering to get to `PVSystem.module_parameters` value (#959 is relevant).

**Comment by mikofski** [comment]
I would look at the datasheets and check the vintages:
* I believe the SunPower SPR-400E-WHT-D is a really old module, most similar to SunPower's Series-E modules like this one: [SPR-E20-435-COM](https://us.sunpower.com/sites/default/files/media-library/data-sheets/sunpower-e-series-commercial-solar-panels-e20-435-com-datasheet-521912-revb_1.pdf) which uses different technology and a smaller wafer than the newer AMAX modules
* The [SunPower SPR-MAX3-400](https://sunpower.maxeon.com/au/sites/default/files/2019-07/max3-400-390-370-au_0.pdf) might be in the SAM CEC module list, but you may not be finding it. I think you should look for "SunPower SPR-A400". This module only has 104 cells (13-rows x 12-columns) because it uses a newer A-Series cell technology. I think the cells are a different size, and also have different characteristics from the older E-Series cells?

I was able to get the AMAX modules like this:
```python
>>> import pvlib
>>> url = "https://raw.githubusercontent.com/NREL/SAM/develop/deploy/libraries/CEC%20Modules.csv"
>>> cec_modules = pvlib.pvsystem.retrieve_sam(name=None, path=url)
>>> sunpower_modules = cec_modules.T[cec_modules.T.index.str.startswith('SunPower')]
>>> sunpower_modules[sunpower_modules.index.str.contains('A400')]

#                        Manufacturer Technology Bifacial     STC  ... gamma_r BIPV           Version        Date
# SunPower_SPR_A400          SunPower  Mono-c-Si        0  399.96  ...  -0.311    N  SAM 2020.2.29 r3  11/25/2020
# SunPower_SPR_A400_COM      SunPower  Mono-c-Si        0  437.33  ...  -0.329    N  SAM 2020.2.29 r3  11/25/2020
# SunPower_SPR_A400_G_AC     SunPower  Mono-c-Si        0  399.96  ...  -0.311    N  SAM 2020.2.29 r3  11/25/2020
# SunPower_SPR_A400_MLSD     SunPower  Mono-c-Si        0  399.96  ...  -0.311    N  SAM 2020.2.29 r3  11/25/2020

[4 rows x 26 columns]
```

**Reply by mikofski** [reply]
Maybe @markcampanelli or someone from SunPower can comment about SP AMAX 400W modules?

**Reply by casaout** [reply]
Thank you @mikofski. I agree that it makes sense to also consider the module's vintages. Since the `*_A400*` ones all have 66 or 72 cells (and `SunPower SPR-MAX3-400` has 104 modules),  I am wondering about the scaling (see the [other thread](https://github.com/pvlib/pvlib-python/discussions/1344#discussioncomment-1806989)). Since the STC is already correct at around 400W I am unsure about balancing. Do you have an idea for approaching that?

| | SunPower SPR-MAX3-400 (target) | SunPower SPR-A400 (closest) | scaled (i.e. /66*104)|
|--|--|--|--|
|number cells | 104 | 66| 104 (nice!)|
|v_oc_ref|75.6V|47.6V|75.0V (nice!)|
|v_mp_ref|65.8V|39.6V|62.4V (nice!)|
|STC|400W|399.96W|630W (how to approach scaling?)|
|I_L_ref|??|10.9258A|?|

**Reply by mikofski** [reply]
Maybe I was wrong? It looks like your target module has a current about 6A (400W / 65.8V = 6.08A) which is closer to the SPR-E20-435-COM modules. The current of the SPR-A400 module is almost 11A, that doesn't seem like the same technology to me. I think the A400 must have bigger cells, which might explain the higher current? Sorry

**Reply by casaout** [reply]
Generally, it should be a newer one as you suggested (since we just bought it this summer). It might be best if I contact SunPower and ask for clarification.
